libs/ml/include/ml/optimization/adagrad_optimizer.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/standard_functions/sqrt.hpp"
+#include "ml/graph.hpp"
+#include "ml/ops/loss_functions/criterion.hpp"
+#include "ml/optimization/optimizer.hpp"
+
+namespace fetch {
+namespace ml {
+
+/**
+ * Adaptive Gradient Algorithm optimizer
+ * i.e. Modified stochastic gradient descent with per-parameter learning rate
+ * @tparam T ArrayType
+ * @tparam C CriterionType
+ */
+template <class T, class C>
+class AdaGradOptimizer : public Optimizer<T, C>
+{
+public:
+  using ArrayType = T;
+  using DataType  = typename ArrayType::Type;
+  using SizeType  = typename ArrayType::SizeType;
+
+  AdaGradOptimizer(std::shared_ptr<Graph<T>> graph, std::string const &input_node_name,
+                   std::string const &output_node_name, DataType const &learning_rate)
+    : Optimizer<T, C>(graph, input_node_name, output_node_name, learning_rate)
+  {
+    auto weights = this->graph_->GetWeights();
+    for (auto &wei : weights)
+    {
+      this->cache_.push_back(ArrayType(wei.shape()));
+    }
+    ResetCache();
+  }
+
+private:
+  void ApplyGradients()
+  {
+
+    std::vector<ArrayType> gradients = this->graph_->GetGradients();
+
+    // Do operation with gradient
+    SizeType i{0};
+    for (auto &grad : gradients)
+    {
+      auto git = grad.begin();
+      auto cit = cache_[i].begin();
+      while (git.is_valid())
+      {
+        *cit += (*git) * (*git);
+        // 1e-8 is added to prevent division by 0
+        *git = -this->learning_rate_ * ((*git) / fetch::math::Sqrt(*cit + DataType{1e-8f}));
```

Comment:
epsilon value should be a member variable and an input parameter to constructor - not hard-coded